File: src/wickrcrypto/swig/node/test/contextTests.js
'use strict'

var expect = require('expect.js')
var wickrcrypto = require('../../../../../node/lib/wickrcrypto')

describe ("Context Tests", function() {
    
    var devinfo
    var identifier
    var ctx

    beforeEach(function() {
        //Generate a context from only device info and an identifier
        devinfo = wickrcrypto.DeviceInfo.gen(Buffer.from("sysid"))
        identifier = wickrcrypto.CryptoEngine.digest(Buffer.from("wickr"), null, wickrcrypto.Digest.sha256())
        var testGeneration = wickrcrypto.ContextGenResult.genNew(devinfo, identifier)
        expect(testGeneration).to.be.a("object")
        ctx = testGeneration.ctx
        expect(ctx.idChain.root.signature).to.eql(testGeneration.ctx.idChain.root.signature)
        expect(ctx).to.be.a("object")
    })

    it ("can be created from individual values", function() {
        var testFromValues = wickrcrypto.Context.fromValues(ctx.devInfo, ctx.idChain, ctx.storageKeys)
        expect(testFromValues).to.be.a("object")
    })

    it ("can be generated", function() {
        
        this.timeout(15000)

        var existingSigKey = wickrcrypto.CryptoEngine.randEcKey(wickrcrypto.ECCurve.p521())

        //Generate a context from an existing signing key, device info, and an identifier
        var testGenerationSig = wickrcrypto.ContextGenResult.genNewWithSigKey(devinfo, existingSigKey, identifier)
        expect(testGenerationSig).to.be.a("object")
        expect(testGenerationSig.ctx.idChain.root.sigKey.pubData).to.eql(existingSigKey.pubData)

        var testKeys = wickrcrypto.RootKeys.gen()

        //Generate a context from existing root keys, device info, and an identifier
        var testGenerationRootKeys = wickrcrypto.ContextGenResult.genWithRootKeys(devinfo, testKeys, identifier)
        expect(testGenerationRootKeys).to.be.a("object")

        expect(testKeys.nodeStorageRoot.keyData).to.eql(testGenerationRootKeys.rootKeys.nodeStorageRoot.keyData)
        expect(testKeys.remoteStorageRoot.keyData).to.eql(testGenerationRootKeys.rootKeys.remoteStorageRoot.keyData)
        expect(testKeys.nodeSignatureRoot.pubData).to.eql(testGenerationRootKeys.rootKeys.nodeSignatureRoot.pubData)

        //Create recovery data so the context can be regenerated with a recovery key
        var recoveryData = testGenerationRootKeys.makeRecovery()
        expect(recoveryData).to.be.a("object")

        //Export the recovery key of a context with a passphrase
        var exportedRecoveryKey = testGenerationRootKeys.exportRecoveryKeyPassphrase(Buffer.from("password"))
        expect(exportedRecoveryKey).to.be.a("object")

        expect(wickrcrypto.ContextGenResult.importRecoveryKeyPassphrase(exportedRecoveryKey, Buffer.from("password2"))).to.not.be.a("object")

        //Import a recovery key with a passphrase
        var importedRecoveryKey = wickrcrypto.ContextGenResult.importRecoveryKeyPassphrase(exportedRecoveryKey, Buffer.from("password"))
        expect(importedRecoveryKey).to.be.a("object")

        expect(importedRecoveryKey.keyData).to.eql(testGenerationRootKeys.recoveryKey.keyData)

        //Regenerate a context with a recovery key, recovery information, device info, and an identifier
        var testRecoveryGeneration = wickrcrypto.ContextGenResult.genWithRecovery(devinfo, recoveryData, importedRecoveryKey, identifier)

        expect(testRecoveryGeneration).to.be.a("object")

        expect(wickrcrypto.ContextGenResult.genWithPassphrase(devinfo, exportedRecoveryKey, Buffer.from("password2"), recoveryData, identifier)).to.not.be.a("object")

        //Regenerate a context with a passphrase, exported recovery key, device info, an identifier, and recovery data
        var testPassphraseGeneration = wickrcrypto.ContextGenResult.genWithPassphrase(devinfo, exportedRecoveryKey, Buffer.from("password"), recoveryData, identifier)

        expect(testPassphraseGeneration).to.be.a("object")
    })

    it ("can perform cipher ops", function() {
        var testData = Buffer.from("test")

        //Cipher local data using the context
        var ciphered = ctx.cipherLocal(testData)
        expect(ciphered).to.be.a("object")

        //Decipher local data using the context
        var deciphered = ctx.decipherLocal(ciphered)
        expect(testData).to.eql(deciphered)

        //Cipher remote data using the context
        ciphered = ctx.cipherRemote(testData)
        expect(ciphered).to.be.a("object")

        //Decipher remote data using the context
        deciphered = ctx.decipherRemote(ciphered)
        expect(testData).to.eql(deciphered)
    })

    it ("can generate keys", function() {
        //Generate ephemeral keypairs to distribute for message sending
        var testKey1 = ctx.ephemeralKeypairGen(1)
        expect(testKey1).to.be.a("object")
        expect(testKey1.identifier).to.eql(1)

        var testKey2 = ctx.ephemeralKeypairGen(2)
        expect(testKey2).to.be.a("object")
        expect(testKey2.identifier).to.eql(2)
        expect(testKey1.ecKey.pubData).to.not.eql(testKey2.ecKey.pubData)
        expect(testKey1.ecKey.priData).to.not.eql(testKey2.ecKey.priData)

        //Erase the public key of the ephemeral keypair
        testKey1.makePublic()
        expect(testKey1.ecKey.priData).to.not.be.a("object")
    })

    it ("can export storage keys", function() {
        this.timeout(15000)
        //Export a storage key with a passphrase
        var exported = ctx.exportStorageKeys(Buffer.from("password"))
        expect(exported).to.be.a("object")

        //Import a storage key with a passphrase
        expect(wickrcrypto.Context.importStorage(exported, Buffer.from("password2"))).to.not.be.a("object")
        var imported = wickrcrypto.Context.importStorage(exported, Buffer.from("password"))
        expect(imported).to.be.a("object")
    })

    it ("can encode and decode messages", function() {
        //Generate a context to 'send' messages to
        var devinfo = wickrcrypto.DeviceInfo.gen(Buffer.from("sysid2"))
        var identifier = wickrcrypto.CryptoEngine.digest(Buffer.from("wickr"), null, wickrcrypto.Digest.sha256())
        var testGeneration = wickrcrypto.ContextGenResult.genNew(devinfo, identifier)
        expect(testGeneration).to.be.a("object")

        var receiverCtx = testGeneration.ctx
        expect(receiverCtx).to.be.a("object")

        var msgKey = receiverCtx.ephemeralKeypairGen(1)
        expect(msgKey).to.be.a("object")

        var receiverNode = wickrcrypto.Node.fromValues(devinfo.msgProtoId, receiverCtx.idChain, msgKey)
        expect(receiverNode).to.be.a("object")

        var message = Buffer.from("wickr")
        var messageType = 1000
        var channel = Buffer.from("12345")
        var ephemerality = wickrcrypto.EphemeralInfo.fromValues(86400, 500)

        //Create metadata for this message
        var meta = wickrcrypto.PacketMeta.fromValues(ephemerality, channel, messageType)
        expect(meta).to.be.a("object")

        //Construct a payload with a message body and metadata
        var payload = wickrcrypto.Payload.fromValues(meta, message)
        expect(payload).to.be.a("object")

        var nodes = wickrcrypto.WickrArray.allocateNode(1)
        nodes.setNode(0,receiverNode)
        expect(nodes.getItemCount()).to.eql(1)

        //Encode the message for a set of node receivers
        var encodeResult = ctx.encodePacket(payload, nodes)
        expect(encodeResult).to.be.a("object")

        //Use the receiver context created above to parse a packet for non decoding purposes
        var parsed = receiverCtx.parsePacketNoDecode(encodeResult.encodedPacket, ctx.idChain)
        expect(parsed).to.be.a("object")
        expect(parsed.parseResult.err).to.eql(wickrcrypto.E_SUCCESS)
        expect(parsed.parseResult.signatureStatus).to.eql(wickrcrypto.PACKET_SIGNATURE_VALID)

        //Use the receiver context created above to parse a packet for decoding purposes
        parsed = receiverCtx.parsePacket(encodeResult.encodedPacket, ctx.idChain)
        expect(parsed).to.be.a("object")
        expect(parsed.parseResult.err).to.eql(wickrcrypto.E_SUCCESS)
        expect(parsed.parseResult.signatureStatus).to.eql(wickrcrypto.PACKET_SIGNATURE_VALID)

        //Decode the packet, and extract the original message
        var decoded = receiverCtx.decodePacket(parsed, msgKey.ecKey)
        expect(decoded).to.be.a("object")
        expect(decoded.err).to.eql(wickrcrypto.E_SUCCESS)
        expect(decoded.decryptedPayload.body).to.eql(message)

    })

    it("can be serialized and deserialized", function() {

        //Serialize and deserialize the context
        var serializedContext = ctx.serialize()
        expect(serializedContext).to.be.a("object")

        var restoredContext = wickrcrypto.Context.fromBuffer(devinfo, serializedContext)
        expect(restoredContext).to.be.a("object")

        //Test some properties to ensure proper wrapping, this is heavily tested in the wickr-crypto-c library tests
        expect(restoredContext.devInfo.msgProtoId).to.eql(devinfo.msgProtoId)
        expect(restoredContext.idChain.root.identifier).to.eql(identifier)

    })

    it("can be exported and imported", function() {

        var password = Buffer.from('password')
        //Serialize and deserialize the context
        var exportedContext = ctx.exportPassphrase(password)
        expect(exportedContext).to.be.a("object")

        var restoredContext = wickrcrypto.Context.importFromBuffer(devinfo, exportedContext, password)
        expect(restoredContext).to.be.a("object")

        //Test some properties to ensure proper wrapping, this is heavily tested in the wickr-crypto-c library tests
        expect(restoredContext.devInfo.msgProtoId).to.eql(devinfo.msgProtoId)
        expect(restoredContext.idChain.root.identifier).to.eql(identifier)

    })

})

File: src/wickrcrypto/swig/java/com/wickr/crypto/tests/ContextTests.java
package com.wickr.crypto.tests;

import static org.junit.Assert.*;
import static org.hamcrest.CoreMatchers.*;
import org.junit.Test;
import com.wickr.crypto.*;
import java.io.*;
import java.util.*;
import org.junit.Before;
import java.math.BigInteger;

public class ContextTests
{
	protected DeviceInfo devinfo;
	protected byte[] identifier;
	protected Context ctx;

	@Before
    public void setUp() throws UnsupportedEncodingException
    {
    	//Generate a context from only device info and an identifier
    	devinfo = DeviceInfo.gen("sysid".getBytes("UTF8"));
        identifier = CryptoEngine.digest("wickr".getBytes("UTF8"), null, Digest.sha256());
        ContextGenResult testGeneration = ContextGenResult.genNew(devinfo, identifier);
		assertNotNull(testGeneration);
		ctx = testGeneration.getCtx();
		assertNotNull(ctx);
    }

    @Test
    public void testContextFromValues() {
		Context testFromValues = Context.fromValues(ctx.getDevInfo(), ctx.getIdChain(), ctx.getStorageKeys());
    	assertNotNull(testFromValues);
	}

    @Test
	public void testContextGeneration() throws UnsupportedEncodingException
	{

		ECKey existingSigKey = CryptoEngine.randEcKey(ECCurve.p521());

		//Generate a context from an existing signing key, device info, and an identifier
		ContextGenResult testGenerationSig = ContextGenResult.genNewWithSigKey(devinfo, existingSigKey, identifier);
		assertNotNull(testGenerationSig);
		assertArrayEquals(testGenerationSig.getCtx().getIdChain().getRoot().getSigKey().getPubData(), existingSigKey.getPubData());

		RootKeys testKeys = RootKeys.gen();

		//Generate a context from existing root keys, device info, and an identifier
		ContextGenResult testGenerationRootKeys = ContextGenResult.genWithRootKeys(devinfo, testKeys, identifier);
		assertNotNull(testGenerationRootKeys);

		assertArrayEquals(testKeys.getNodeStorageRoot().getKeyData(), testGenerationRootKeys.getRootKeys().getNodeStorageRoot().getKeyData());
		assertArrayEquals(testKeys.getRemoteStorageRoot().getKeyData(), testGenerationRootKeys.getRootKeys().getRemoteStorageRoot().getKeyData());
		assertArrayEquals(testKeys.getNodeSignatureRoot().getPubData(), testGenerationRootKeys.getRootKeys().getNodeSignatureRoot().getPubData());

		//Create recovery data so the context can be regenerated with a recovery key
		byte[] recoveryData = testGenerationRootKeys.makeRecovery();
		assertNotNull(recoveryData);

		//Export the recovery key of a context with a passphrase
		byte[] exportedRecoveryKey = testGenerationRootKeys.exportRecoveryKeyPassphrase("password".getBytes("UTF8"));
		assertNotNull(exportedRecoveryKey);

		assertNull(ContextGenResult.importRecoveryKeyPassphrase(exportedRecoveryKey, "password2".getBytes("UTF8")));

		//Import a recovery key with a passphrase
		CipherKey importedRecoveryKey = ContextGenResult.importRecoveryKeyPassphrase(exportedRecoveryKey, "password".getBytes("UTF8"));
		assertNotNull(importedRecoveryKey);

		assertArrayEquals(importedRecoveryKey.getKeyData(), testGenerationRootKeys.getRecoveryKey().getKeyData());

		//Regenerate a context with a recovery key, recovery information, device info, and an identifier
		ContextGenResult testRecoveryGeneration = ContextGenResult.genWithRecovery(devinfo, recoveryData, importedRecoveryKey, identifier);

		assertNotNull(testRecoveryGeneration);

		assertNull(ContextGenResult.genWithPassphrase(devinfo, exportedRecoveryKey, "password2".getBytes("UTF8"), recoveryData, identifier));

		//Regenerate a context with a passphrase, exported recovery key, device info, an identifier, and recovery data
		ContextGenResult testPassphraseGeneration = ContextGenResult.genWithPassphrase(devinfo, exportedRecoveryKey, "password".getBytes("UTF8"), recoveryData, identifier);

		assertNotNull(testPassphraseGeneration);
	}

	@Test
	public void testContextCipherOps() throws UnsupportedEncodingException
	{
		byte[] testData = "test".getBytes("UTF8");

		//Cipher local data using the context
		CipherResult ciphered = ctx.cipherLocal(testData);
		assertNotNull(ciphered);

		//Decipher local data using the context
		byte[] deciphered = ctx.decipherLocal(ciphered);
		assertArrayEquals(testData, deciphered);

		//Cipher remote data using the context
		ciphered = ctx.cipherRemote(testData);
		assertNotNull(ciphered);

		//Decipher remote data using the context
		deciphered = ctx.decipherRemote(ciphered);
		assertArrayEquals(testData, deciphered);
	}

	@Test
	public void testContextGenKeys() throws UnsupportedEncodingException
	{
		//Generate ephemeral keypairs to distribute for message sending
		EphemeralKeypair testKey1 = ctx.ephemeralKeypairGen(new BigInteger("1"));
		assertNotNull(testKey1);
		assertEquals(testKey1.getIdentifier().toString(), "1");

		EphemeralKeypair testKey2 = ctx.ephemeralKeypairGen(new BigInteger("2"));
		assertNotNull(testKey2);
		assertEquals(testKey2.getIdentifier().toString(), "2");
		assertNotEquals(testKey1.getEcKey().getPubData(), testKey2.getEcKey().getPubData());
		assertNotEquals(testKey1.getEcKey().getPriData(), testKey2.getEcKey().getPriData());

		//Erase the public key of the ephemeral keypair
		testKey1.makePublic();
		assertNull(testKey1.getEcKey().getPriData());

	}

	@Test
	public void testStorageKeys() throws UnsupportedEncodingException
	 {
		//Export a storage key with a passphrase
		byte[] exported = ctx.exportStorageKeys("password".getBytes("UTF8"));
		assertNotNull(exported);

		//Import a storage key with a passphrase
		assertNull(ctx.importStorage(exported, "password2".getBytes("UTF8")));
		StorageKeys imported = ctx.importStorage(exported, "password".getBytes("UTF8"));
		assertNotNull(imported);
	}

	@Test
	public void testGenerateMessages() throws UnsupportedEncodingException
	{

		//Generate a context to 'send' messages to
		DeviceInfo devinfo = DeviceInfo.gen("sysid2".getBytes("UTF8"));
        byte[] identifier = CryptoEngine.digest("wickr2".getBytes("UTF8"), null, Digest.sha256());
        ContextGenResult testGeneration = ContextGenResult.genNew(devinfo, identifier);
		assertNotNull(testGeneration);

		Context receiverCtx = testGeneration.getCtx();
		assertNotNull(ctx);

		EphemeralKeypair msgKey = receiverCtx.ephemeralKeypairGen(new BigInteger("1"));
		assertNotNull(msgKey);

		Node receiverNode = Node.fromValues(devinfo.getMsgProtoId(), receiverCtx.getIdChain(), msgKey);
		assertNotNull(receiverNode);

		byte[] message = "Wickr".getBytes("UTF8");
		int messageType = 1000;
		byte[] channel = "12345".getBytes("UTF8");
		EphemeralInfo ephemerality = EphemeralInfo.fromValues(new BigInteger("86400"), new BigInteger("500"));

		//Create metadata for this message
		PacketMeta meta = PacketMeta.fromValues(ephemerality, channel, messageType);
		assertNotNull(meta);

		//Construct a payload with a message body and metadata
		Payload payload = Payload.fromValues(meta, message);
		assertNotNull(payload);

		WickrArray nodes = WickrArray.allocateNode(1);
		nodes.setNode(0, receiverNode);

		//Encode the message for a set of node receivers
		EncoderResult encodeResult = ctx.encodePacket(payload, nodes);
		assertNotNull(encodeResult);

		//Use the receiver context created above to parse a packet for non decoding purposes
		ContextParseResult parsed = receiverCtx.parsePacketNoDecode(encodeResult.getPacket().serialize(), ctx.getIdChain());
		assertNotNull(parsed);
		assertEquals(parsed.getParseResult().getErr(), DecodeError.E_SUCCESS);
		assertEquals(parsed.getParseResult().getSignatureStatus(), PacketSignatureStatus.PACKET_SIGNATURE_VALID);

		//Use the receiver context created above to parse a packet for decoding purposes
		parsed = receiverCtx.parsePacket(encodeResult.getPacket().serialize(), ctx.getIdChain());
		assertNotNull(parsed);
		assertEquals(parsed.getParseResult().getErr(), DecodeError.E_SUCCESS);
		assertEquals(parsed.getParseResult().getSignatureStatus(), PacketSignatureStatus.PACKET_SIGNATURE_VALID);

		//Decode the packet, and extract the original message
		ContextDecodeResult decoded = receiverCtx.decodePacket(parsed, msgKey.getEcKey());
		assertNotNull(decoded);
		assertEquals(decoded.getErr(), DecodeError.E_SUCCESS);
		assertArrayEquals(decoded.getDecryptedPayload().getBody(), message);
	}

	@Test
	public void testContextSerialization() {

		//Serialize and deserialize the context
		byte[] serializedContext = ctx.serialize();
		assertNotNull(serializedContext);
		assertTrue(serializedContext.length > 0);

		Context restoredContext = Context.fromBuffer(devinfo, serializedContext);
		assertNotNull(restoredContext);

		//Test some properties to ensure proper wrapping, this is heavily tested in the wickr-crypto-c library tests
		assertArrayEquals(restoredContext.getDevInfo().getMsgProtoId(), devinfo.getMsgProtoId());
		assertArrayEquals(restoredContext.getIdChain().getRoot().getIdentifier(), identifier);

	}

	@Test
	public void testContextExport() throws UnsupportedEncodingException {

		byte[] password = "password".getBytes("UTF8");
		//Serialize and deserialize the context
		byte[] exportedContext = ctx.exportPassphrase(password);
		assertNotNull(exportedContext);
		assertTrue(exportedContext.length > 0);

		Context restoredContext = Context.importFromBuffer(devinfo, exportedContext, password);
		assertNotNull(restoredContext);

		//Test some properties to ensure proper wrapping, this is heavily tested in the wickr-crypto-c library tests
		assertArrayEquals(restoredContext.getDevInfo().getMsgProtoId(), devinfo.getMsgProtoId());
		assertArrayEquals(restoredContext.getIdChain().getRoot().getIdentifier(), identifier);

	}
}